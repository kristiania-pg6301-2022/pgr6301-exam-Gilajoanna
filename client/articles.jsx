import * as React from "react";
import { Link, Route, Routes } from "react-router-dom";
import { useLoader } from "./globals/useLoader";
import { fetchJSON } from "./globals/fetchJSON";
import { AddNewArticle } from "./addNewArticle";

function ArticleItem({ article: { title, category, content, author } }) {
  return (
    <div>
      <h2>{title}</h2>
      <div>
        <strong>{category}</strong>
      </div>
      <div>
        Author: <i>{author}</i>
      </div>
      <div>{content}</div>
    </div>
  );
}

export function ListArticles({ listAllArticles }) {
  const { loading, data, error } = useLoader(listAllArticles);

  if (loading) {
    return <div>Please wait..</div>;
  }

  if (error) {
    return (
      <div>
        <h1>An error occurred</h1>
        <div id="error-message">Something went wrong. Please try again.</div>
      </div>
    );
  }

  return (
    <div>
      <h1>List of articles</h1>

      {data.map((article) => (
        <ArticleItem key={article.title} article={article} />
      ))}
    </div>
  );
}

export function Articles() {
  async function listAllArticles() {
    return await fetchJSON("/api/articles");
  }

  return (
    <Routes>
      <Route
        path={""}
        element={<ListArticles listAllArticles={listAllArticles} />}
      />
      <Route path={"new"} element={<AddNewArticle />} />
    </Routes>
  );
}
