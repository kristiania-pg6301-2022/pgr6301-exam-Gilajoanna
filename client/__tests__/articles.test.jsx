import * as ReactDOM from "react-dom";
import * as React from "react";
import { ListArticles } from "../articles.jsx";
import { act } from "react-dom/test-utils";

describe("List articles component", () => {
  it("renders ListArticle", () => {
    const element = document.createElement("div");
    ReactDOM.render(<ListArticles />, element);
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("shows list of articles", async () => {
    const article = [{ title: "Article" }];
    const element = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <ListArticles listAllArticles={() => article} />,
        element
      );
    });

    expect(
      Array.from(element.querySelectorAll("h2")).map((e) => e.innerHTML)
    ).toEqual(["Article"]);

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("shows error", async () => {
    const element = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <ListArticles
          listAllArticles={() => {
            throw new Error("Something went wrong. Please try again.");
          }}
        />,
        element
      );
    });
    expect(element.querySelector("#error-message").innerHTML).toEqual(
      "Something went wrong. Please try again."
    );
    expect(element.innerHTML).toMatchSnapshot();
  });
});
