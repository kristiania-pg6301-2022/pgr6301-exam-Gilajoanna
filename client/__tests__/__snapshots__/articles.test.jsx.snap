// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List articles component renders ListArticle 1`] = `"<div>Please wait..</div>"`;

exports[`List articles component shows error in article list 1`] = `"<div><h1>An error occurred</h1><div id=\\"error-message\\">Something went wrong. Please try again.</div></div>"`;

exports[`List articles component shows error in aside article list 1`] = `"<div><h1>An error occurred</h1><div id=\\"error-message\\">Something went wrong. Please try again.</div></div>"`;

exports[`List articles component shows list of articles 1`] = `"<div><div><h2>Article</h2><div><strong></strong></div><div>Writer: <i></i></div><div></div><button type=\\"submit\\">Delete Article</button></div></div>"`;
